testing/upload_validation.ts

```diff
@@ -227,6 +228,120 @@ export function validateSyncTableSchema(
   );
 }
 
+/**
+ * Returns a map of sync table names to their child sync table names, or undefined if the hierarchy is invalid.
+ * Example valid return: { Parent: 'Child' }
+ * {} is also a valid result, when there are no sync tables, or no parent relationships.
+ * @hidden
+ */
+export function validateCrawlHierarchy(
+  pack: PackVersionDefinition,
+  context?: z.RefinementCtx,
+): Record<string, string[]> | undefined {
+  const result: Record<string, string[]> = {};
+  if (!pack.syncTables) {
+    return result;
+  }
+  const syncTableSchemasByName: Record<string, ObjectSchema<any, any>> = {};
+  for (const syncTable of pack.syncTables) {
+    syncTableSchemasByName[syncTable.name] = syncTable.schema;
+  }
+  for (const [index, syncTable] of pack.syncTables.entries()) {
+    let parent: string | undefined;
+    for (const param of syncTable.getter.parameters) {
+      if (!param.crawlStrategy) {
+        continue;
+      }
+      if (param.crawlStrategy.parentTable) {
+        const {tableName, propertyKey} = param.crawlStrategy.parentTable;
+
+        const tableSchema = syncTableSchemasByName[tableName];
+        if (!tableSchema) {
+          context?.addIssue({
+            code: z.ZodIssueCode.custom,
+            path: ['syncTables', index],
+            message: `Sync table ${syncTable.name} expects parent table ${tableName} to exist.`,
+          });
+          return undefined;
+        }
+
+        const property = tableSchema.properties[propertyKey];
+        if (!property) {
+          context?.addIssue({
+            code: z.ZodIssueCode.custom,
+            path: ['syncTables', index],
+            message: `Sync table ${syncTable.name} expects parent table ${tableName}'s schema to have the property ${propertyKey}.`,
+          });
+          return undefined;
+        }
+
+        // TODO(patrick): Validate the types match
+
+        // We only allow one parent per table.
+        if (parent && parent !== tableName) {
+          context?.addIssue({
+            code: z.ZodIssueCode.custom,
+            path: ['syncTables', index],
```

Comment:
This one is probably ok since both crawlStrategies are valid on their own but it's the presence of both that creates the violation.

Reply:
I'll just add 'parameters' without an index.